Hot Reloading (true story)

AlbertVVila opened the pull request:
# Le Hot Feature

![imagen](https://user-images.githubusercontent.com/25633904/61368046-4c91f600-a88d-11e9-9831-38fa88834652.png)

## What is this?
 A way of editing scripts while the Engine is running that speed ups the gameplay programming.
## Okey but what do I have to do?
You have to open the Engine not with Visual Studio but instead you have to use the generated EXE.
1. Compile engine+scripts
2. Open the folder where the EXE has been created (for instance if you choose Release, it's under Engine/Release)
3. Copy this in to the Game/ folder
4. Launch the Engine from there
5. Edit scripts in visual studio and compile them whenever you want. (One script compiled at a time is supported).
6. Once it is compiled it will be applied while the Engine is running, depens on compiling times (usually less 0.5 seconds or less).

## Does it work well?
- **Engine Running**: If you are not in Play mode it is guaranteed that changes will be applied correctly.
- **Engine Playing**: If you are playing it depens on the script. For small and isolated scripts it will work for most cases. For huge and important scripts it will probably work too but it can create some small bugs.
- **Game Build**: It is disabled for GameBuild and GameBuild Debug. It could be enabled in the future if we needed to.
## What to do to enable it?
- In Configuration on the Engine there is a tab for scripts where you can toggle the hot reloading. It is enabled by default.
- Once this PR is approved I will upload a new TemplateScript adapter for hotreloading (hopefully the last TemplateScript version ever).
## Nice, how does it really work?
When you compile a script while the engine is running I create a copy of the script with the _.hot_ extension. Then the Engine detects it and unloads the given script, if it can, serializing the information of the script and the gameobject associated. In case the script could'nt be serialized because of dependencies, it unloads all the scripts and reloads all of them afterwards with the new compiled script.

## Couple of extra things
- Optimized Script detection, now it's done in a separate thread (decreases frame time by 3-4ms in Release mode).
- Changed a method in FSystem to rename files changing also their extension.

## Le Test
- Simply follow the instructions above and try to modify a script, compile it and see if that changes are correctly applied without the Engine crashing. Don't push it too much, the feature is not ready for extreme cases.
- It is highly recommended to compile modified scripts and not recompile them. The difference is that if you recompile, all the script references will be recompiled aswell and thus increases the possible issues regarding hotreloading.
- In case you try to hotreloading while playing and it doesn't work as you thought, try to test it again outside of play mode to check if it is actually a bug in the hot reload feature or a hot reloaded script that doesn't behave correctly while playing.

AlfonsoMestres commented on Source/ModuleScript.cpp:
```diff
@@ -213,51 +254,228 @@ bool ModuleScript::RemoveScript(Script* script, const std::string& name)
+void ModuleScript::UpdateScript(std::string scriptName)
+{
+	hotJson = new JSON();
+	for (unsigned i = 0; i < componentsScript.size(); ++i)
```
unsigned i = 0u

AlbertVVila replied:
0u mama

AlbertVVila replied:
done

AlfonsoMestres replied:
RIP

AlfonsoMestres commented on Source/ModuleScript.cpp:
```diff
@@ -213,51 +254,228 @@ bool ModuleScript::RemoveScript(Script* script, const std::string& name)
+	}
+	else //Reload All
+	{
+		for (unsigned i = 0; i < componentsScript.size(); ++i)
```
unsigned i = 0u

AlbertVVila replied:
👍

AlbertVVila commented on Source/ModuleFileSystem.cpp:
```diff
@@ -409,9 +409,17 @@ bool ModuleFileSystem::Rename(const char* route, const char* file, const char* n
+		}
 		Copy(route, file, route, (newName + extension).c_str());
-		success = Delete(filepath.c_str());
+		success = Delete(filepath.c_str()); //TODO: only if it doesn't overwrite another file
```
I think it's important to keep that TODO in mind 😃

AlfonsoMestres replied:
Why is not implemented if it's important?

AlbertVVila replied:
It's important to remember it, but for now I don't think it's necesary 🙆‍♂

AlbertVVila commented:
> 
> 
> When recompiling a script that has multiple references to other scripts it doesn't seem to be working. multiple .hot files are being created in the Scripts folder but the changes do not apply to them.
> 
> It should support having references to other scripts.
> 
> Le review

Corrected those bugs, and modified the description to add further information for testing purposes
